cpp/doc/DEVELOPER_GUIDE.md

```diff
@@ -0,0 +1,739 @@
+`std::vector<T>`.
+
+Alternatively, C++17
+[structured binding](https://en.cppreference.com/w/cpp/language/structured_binding) may be used to
+disaggregate multiple return values:
+
+```c++
+auto [out0, out1] = cudf::return_two_outputs();
+```
+
+Note that the compiler might not support capturing aliases defined in a structured binding
+in a lambda. One may work around this by using a capture with an initializer instead:
+
+```c++
+auto [out0, out1] = cudf::return_two_outputs();
+
+// Direct capture of alias from structured binding might fail with:
+// "error: structured binding cannot be captured"
+// auto foo = [out0]() {...};
+
+// Use an initializing capture:
+auto foo = [&out0 = out0] {
+  // Use out0 to compute something.
+  // ...
+};
+```
+
+
+## Streams
+
+CUDA streams are not yet exposed in public column-based libcuspatial APIs. header-only libcuspatial
+APIs that execute GPU work or allocate GPU memory should take a stream parameter.
+
+In order to ease the transition to future use of streams in the public column-based API, all
+libcuspatial APIs that allocate device memory or execute GPU work (including kernels, 
+Thrust algorithms, or anything that can take a stream) should be implemented using asynchronous APIs
+on the default stream (e.g., stream 0).
+
+The recommended pattern for doing this is to make the definition of the external API invoke an
+internal API in the `detail` namespace. The internal `detail` API has the same parameters as the
+public API, plus a `rmm::cuda_stream_view` parameter at the end with no default value. If the
+detail API also accepts a memory resource parameter, the stream parameter should be ideally placed
+just *before* the memory resource. The public API will call the detail API and provide
+`rmm::cuda_stream_default`. The implementation should be wholly contained in the `detail` API
+definition and use only asynchronous versions of CUDA APIs with the stream parameter.
+
+In order to make the `detail` API callable from other libcudf functions, it should be exposed in a
+header placed in the `cudf/cpp/include/detail/` directory.
+
+For example:
+
+```c++
+// cpp/include/cuspatial/header.hpp
+void external_function(...);
+
+// cpp/include/cuspatial/detail/header.hpp
+namespace detail{
+void external_function(..., rmm::cuda_stream_view stream)
+} // namespace detail
+
+// cuspatial/src/implementation.cpp
+namespace detail{
+    // Use the stream parameter in the detail implementation.
+    void external_function(..., rmm::cuda_stream_view stream){
+        // Implementation uses the stream with async APIs.
+        rmm::device_buffer buff(...,stream);
+        CUSPATIAL_CUDA_TRY(cudaMemcpyAsync(...,stream.value()));
+        kernel<<<..., stream>>>(...);
+        thrust::algorithm(rmm::exec_policy(stream), ...);
+    }
+} // namespace detail
+
+void external_function(...){
+    CUDF_FUNC_RANGE(); // Generates an NVTX range for the lifetime of this function.
```

Comment:
We don't have this in the cuspatial domain yet. Perhaps create `CUSPATIAL_FUNC_RANGE()`?

Reply:
Added.